c++/src/kj/async-inl.h

```diff
@@ -2063,8 +2063,39 @@ concept NoWaitScope = !isSameType<Decay<T>, WaitScope>();
 // Define a Concept to use in our `coroutine_traits` specialization to validate allowable coroutine
 // parameter types.
 
+struct DisallowedWithinCoroutine {};
+
+template <typename T, typename = int>
+struct isDisallowedInCoroutine {
+  static constexpr bool value = false;
+};
+
+template <typename T>
+struct isDisallowedInCoroutine<T,
+    decltype((void) ((instance<T>()))->__kj_disallowed_in_coroutine, 0)> {
+  static constexpr bool value = true;
+};
+
+template <typename T>
+struct isDisallowedInCoroutine<T, decltype((void) Decay<T>::__kj_disallowed_in_coroutine, 0)> {
+  static constexpr bool value = true;
+};
+
 }  // namespace kj::_
 
+#define KJ_DISALLOW_IN_COROUTINE() \
+  void* __kj_disallowed_in_coroutine = nullptr; \
```

Comment:
This declares a new instance member of every type it is used on, increasing the size by 8 bytes. I think it would be better to use a typedef.

```suggestion
  using _kj_DissalowedInCoroutine = void; \
```

I think this could be detected like:

```
template <typename T>
struct IsDisallowedInCoroutine<T, kj::Decay<T>::_kj_DissalowedInCoroutine> {
```

and for pointers:

```
template <typename T>
struct IsDisallowedInCoroutine<T*, T::_kj_DissalowedInCoroutine> {
```

Though I admit my mental model of SFINAE on types is incomplete.